File: agents/browser/src/otlp.ts
// OTLP/JSON construction (docs/contracts/rum.md §2, semantic-conventions.md §10).
//
// The SDK builds OTLP by hand instead of depending on an OpenTelemetry package. That is the difference
// between a few kilobytes and a few hundred: the browser only ever produces spans of five known shapes, and
// the general SDK's exporters, context manager and processors buy nothing here.

import { spanId, traceId } from './ids.js';
import { VERSION } from './version.js';

/** RUM event kinds (`openlog.rum.event`). */
export type RumEvent = 'page_view' | 'vital' | 'error' | 'resource' | 'custom';

export interface AnyAttr {
  [key: string]: string | number | boolean | undefined | null;
}

export interface SpanInput {
  name: string;
  event: RumEvent;
  /** Milliseconds since the epoch. */
  startMs: number;
  /** Duration in milliseconds; 0 for a point-in-time event. */
  durationMs?: number;
  traceId?: string;
  spanId?: string;
  parentSpanId?: string;
  attributes?: AnyAttr;
  exception?: { type: string; message: string; stacktrace: string };
  /** OTLP status code; 2 = ERROR. The server forces this for error events anyway. */
  error?: boolean;
}

interface OtlpKeyValue {
  key: string;
  value: { stringValue: string };
}

interface OtlpEvent {
  timeUnixNano: string;
  name: string;
  attributes: OtlpKeyValue[];
}

export interface OtlpSpan {
  traceId: string;
  spanId: string;
  parentSpanId?: string;
  name: string;
  kind: number;
  startTimeUnixNano: string;
  endTimeUnixNano: string;
  attributes: OtlpKeyValue[];
  events?: OtlpEvent[];
  status?: { code: number };
}

const NANOS = '000000';

function nano(ms: number): string {
  // Milliseconds are integers here; string concatenation avoids BigInt and the 2^53 rounding that
  // `ms * 1e6` would introduce.
  return `${Math.max(0, Math.round(ms))}${NANOS}`;
}

/** Every attribute travels as a string: the backend stores span attributes as Map(String, String). */
function attrs(a: AnyAttr | undefined): OtlpKeyValue[] {
  const out: OtlpKeyValue[] = [];
  if (!a) return out;
  for (const key of Object.keys(a)) {
    const v = a[key];
    if (v === undefined || v === null || v === '') continue;
    out.push({ key, value: { stringValue: String(v) } });
  }
  return out;
}

export function buildSpan(input: SpanInput): OtlpSpan {
  const start = input.startMs;
  const end = start + (input.durationMs ?? 0);
  const span: OtlpSpan = {
    traceId: input.traceId ?? traceId(),
    spanId: input.spanId ?? spanId(),
    name: input.name,
    // INTERNAL for page views, vitals and errors; CLIENT for requests. The server assigns this too, so the
    // value here only matters for readability if it ever changes.
    kind: input.event === 'resource' ? 3 : 1,
    startTimeUnixNano: nano(start),
    endTimeUnixNano: nano(end),
    attributes: attrs(input.attributes),
  };
  if (input.parentSpanId) span.parentSpanId = input.parentSpanId;
  if (input.exception) {
    span.events = [
      {
        timeUnixNano: nano(end),
        name: 'exception',
        attributes: attrs({
          'exception.type': input.exception.type,
          'exception.message': input.exception.message,
          'exception.stacktrace': input.exception.stacktrace,
        }),
      },
    ];
  }
  if (input.error) span.status = { code: 2 };
  return span;
}

export interface ResourceInfo {
  serviceName: string;
  environment: string;
}

/**
 * The OTLP/JSON body of one batch. The resource is sent for completeness and debuggability, but the server
 * rebuilds it from the key (rum.Sanitize): nothing here is trusted.
 */
export function buildPayload(spans: OtlpSpan[], info: ResourceInfo): string {
  const resource: AnyAttr = {
    'service.name': info.serviceName,
    'deployment.environment.name': info.environment,
    'telemetry.sdk.name': 'openlog-browser',
    'telemetry.sdk.language': 'webjs',
    'telemetry.sdk.version': VERSION,
  };
  const nav = typeof navigator !== 'undefined' ? navigator : undefined;
  if (nav) {
    resource['user_agent.original'] = nav.userAgent;
    resource['browser.language'] = nav.language;
    const data = (nav as Navigator & { userAgentData?: { mobile?: boolean; platform?: string } }).userAgentData;
    if (data) {
      if (typeof data.mobile === 'boolean') resource['browser.mobile'] = data.mobile;
      if (data.platform) resource['browser.platform'] = data.platform;
    }
  }
  return JSON.stringify({
    resourceSpans: [
      {
        resource: { attributes: attrs(resource) },
        scopeSpans: [{ scope: { name: 'openlog-browser', version: VERSION }, spans }],
      },
    ],
  });
}

File: agents/kotlin/src/main/kotlin/io/github/onuragtas/openlog/Otlp.kt
package io.github.onuragtas.openlog

// OTLP/JSON encoding (docs/contracts/mobile-agent.md §3).

internal object RumEvent {
    const val SCREEN = "page_view"
    const val VITAL = "vital"
    const val ERROR = "error"
    const val RESOURCE = "resource"
    const val CUSTOM = "custom"
}

/** Nanoseconds as a string: milliseconds are integers here, and concatenation avoids the rounding that
 *  multiplying by 1e6 introduces past 2^53. */
internal fun nanoString(ms: Long): String = "${if (ms < 0) 0 else ms}000000"

/** Every attribute travels as a string: the backend stores span attributes as Map(String, String).
 *  Ordered pairs rather than a map so the cap removes the last ones rather than an arbitrary set. */
internal fun encodeAttributes(attrs: List<Pair<String, Any?>>): List<Map<String, Any>> {
    val out = ArrayList<Map<String, Any>>()
    for ((key, value) in attrs) {
        if (value == null) continue
        val s = value as? String ?: value.toString()
        if (s.isEmpty()) continue // an empty attribute is not a value, and the server drops it anyway
        if (out.size >= Limits.MAX_ATTRIBUTES_PER_SPAN) break
        val limit = if (key == "url.full") Limits.MAX_URL_BYTES else Limits.MAX_ATTR_VALUE_BYTES
        out.add(mapOf("key" to key, "value" to mapOf("stringValue" to truncate(s, limit))))
    }
    return out
}

internal fun buildSpan(
    name: String,
    event: String,
    traceId: String,
    spanId: String,
    startMs: Long,
    durationMs: Long = 0,
    parentSpanId: String? = null,
    attributes: List<Pair<String, Any?>> = emptyList(),
    isError: Boolean = false,
    exception: Triple<String, String, String>? = null,
): Map<String, Any?> {
    val span = LinkedHashMap<String, Any?>()
    span["traceId"] = traceId
    span["spanId"] = spanId
    if (parentSpanId != null) span["parentSpanId"] = parentSpanId
    span["name"] = truncate(name, Limits.MAX_NAME_BYTES)
    // INTERNAL for screens, vitals and errors; CLIENT for requests, like the other clients.
    span["kind"] = if (event == RumEvent.RESOURCE) 3 else 1
    span["startTimeUnixNano"] = nanoString(startMs)
    span["endTimeUnixNano"] = nanoString(startMs + if (durationMs < 0) 0 else durationMs)
    span["attributes"] = encodeAttributes(listOf("openlog.rum.event" to event) + attributes)
    if (exception != null) {
        val (type, message, stack) = exception
        span["events"] = listOf(
            mapOf(
                "timeUnixNano" to nanoString(startMs),
                "name" to "exception",
                "attributes" to encodeAttributes(
                    listOf(
                        "exception.type" to type,
                        "exception.message" to truncate(message, Limits.MAX_MESSAGE_BYTES),
                        "exception.stacktrace" to truncate(stack, Limits.MAX_STACK_BYTES),
                    )
                ),
            )
        )
    }
    if (isError) span["status"] = mapOf("code" to 2)
    return span
}

/** The request body: one resource with its spans. */
internal fun buildPayload(spans: List<Map<String, Any?>>, resource: List<Pair<String, Any?>>): Map<String, Any?> =
    mapOf(
        "resourceSpans" to listOf(
            mapOf(
                "resource" to mapOf("attributes" to encodeAttributes(resource)),
                "scopeSpans" to listOf(mapOf("spans" to spans)),
            )
        )
    )
